File: leetcode/Python/recover-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        dq , v, idx  = collections.deque(), list(), 0
        def traverse(node):
            nonlocal idx
            if node is None:
                return
            traverse(node.left)
            node.val, idx = v[idx], idx + 1
            traverse(node.right)
        dq.append(root)
        while dq:
            cur = dq.popleft()
            v.append(cur.val)
            if cur.left:
                dq.append(cur.left)
            if cur.right:
                dq.append(cur.right)
        v.sort()
        traverse(root)

File: leetcode/C++/recover-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> v;
    int idx;
    void traverse(TreeNode* node){
        if(node==nullptr) return;
        traverse(node->left);
        node->val=v[idx++];
        traverse(node->right);
    }
    void recoverTree(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        TreeNode *cur = nullptr;
        idx = 0;
        while(!q.empty()){
            cur=q.front(); q.pop();
            v.push_back(cur->val);
            if(cur->left) q.push(cur->left);
            if(cur->right) q.push(cur->right);
        }
        sort(v.begin(), v.end());
        traverse(root);
    }
};